lib/segmentation.cpp

```diff
@@ -205,4 +206,318 @@ normalized_dtw_distance(const TimeseriesDistanceFunctionData& data)
+    }
+
+    if (right_array_left_index == right_array_right_index) {
+      max_right_array_num = 0;
+    } else {
+      max_right_array =
+        std::max_element(right_array_left_index, right_array_right_index);
+      max_right_array_num = *max_right_array;
+    }
+
+    if (current_distance < order) {
+      distance_range =
+        order - std::distance(left_array_left_index, left_array_right_index);
+      if (distance_range == order)
+        max_left_array_num =
+          *(std::max_element((score.cend() - distance_range), score.cend()));
+      else
+        max_left_array_num = std::max(
+          *max_left_array,
+          *(std::max_element((score.cend() - distance_range), score.cend())));
+    }
+
+    else if (current_distance > n - order - 1) {
+      distance_range =
+        order - std::distance(right_array_left_index, right_array_right_index);
+      if (distance_range == order)
+        max_right_array_num =
+          *(std::max_element(score.cbegin(), score.cbegin() + distance_range));
+      else
+        max_right_array_num = std::max(
+          *max_right_array,
+          *(std::max_element(score.cbegin(), score.cbegin() + distance_range)));
+    }
+
+    if (*current > max_left_array_num && *current > max_right_array_num) {
+      auto item_index = current - score.cbegin();
+      result.push_back(item_index);
+      // find_max = true;
+      if (current_distance + order + 1 < n) {
+        current = current + order + 1;
+        current_distance = current_distance + order + 1;
+      } else
+        break;
+    }
+
+    else {
+      // find_max = false;
+      if (current_distance + 1 < n) {
+        current++;
+        current_distance++;
+      } else
+        break;
+    }
+
+    if (current == score.end() - 1) {
+      right_array_left_index = score.cend();
+    } else {
+      right_array_left_index = current + 1;
+    }
+
+    if (current_distance >= n - order - 1) {
+      right_array_right_index = score.cend();
+    } else {
+      right_array_right_index = current + order + 1;
+    }
+
+    left_array_right_index = current;
+    if (current_distance < order)
+      left_array_left_index = score.begin();
+    else
+      left_array_left_index = current - order;
+  }
+  return result;
+}
+
+double
+cost_L1_error(EigenTimeSeriesConstRef signal,
+              IndexType start,
+              IndexType end,
+              IndexType min_size)
+{ // the function calculate error with cost function "l1"
+
+  // if (end - start < min_size){ // exeption
+  //   //std::cout << "End - Start < min_size"<<endl;
+  // }
+  std::vector<double> signal_subvector(signal.begin() + start,
+                                       signal.begin() + end);
+
+  double median = median_calculation(signal_subvector);
+  double sum_result = 0.0;
+
+  for (auto& element : signal_subvector) {
+    element =
+      std::abs(element - median); // todo we can delete writing to vector
+    sum_result += element;
+  }
```

Comment:
This loop could be written more modern using `std::accumulate`, but I have no super strong preference towards it: https://godbolt.org/z/v8ETxvM57